### PR TITLE
#1448: Cleanup of stale QA jobs.

### PR DESCRIPTION
**Pull request by paulodamaso**
For #1448:
- Corrected `delete_stale_jobs.groovy` for cleaning stale QA jobs
- Added bundle test for this condition

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1572](https://github.com/zerocracy/farm/issues/1572) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1572](https://github.com/zerocracy/farm/issues/1572) is assigned to @krzyk[/z](https://www.0crat.com/u/krzyk), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/2210504855); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_before.groovy [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import com.zerocracy.pmo.Agenda
+import com.zerocracy.pmo.Pmo
+import com.zerocracy.pmo.Projects
+
+def exec(Project project, XML xml) {
+  Farm farm = binding.variables.farm
+  new Projects(new Pmo(farm), 'test').bootstrap().add('TESTPROJECT')
+  new Agenda(farm, 'test').bootstrap().add(farm.find("@id='TESTPROJECT'")[0], 'gh:test/test#1', 'QA')
+  new Agenda(farm, 'test').bootstrap().add(farm.find("@id='TESTPROJECT'")[0], 'gh:test/test#2', 'QA')
+  new Reviews(farm.find("@id='TESTPROJECT'")[0]).bootstrap().add('gh:test/test#2','test', 'g4s8', new Cash.S('$10'), 30, new Cash.S('$0'))
```
Comment (krzyk) [inline src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_before.groovy]: @paulodamaso this line looks quite long

File: src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_before.groovy [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import com.zerocracy.cash.Cash
+import com.zerocracy.pm.qa.Reviews
+import com.zerocracy.pmo.Agenda
+import com.zerocracy.pmo.Pmo
+import com.zerocracy.pmo.Projects
+
+def exec(Project project, XML xml) {
+  Farm farm = binding.variables.farm
+  new Projects(new Pmo(farm), 'test').bootstrap().add('TESTPROJECT')
+  new Agenda(farm, 'test').bootstrap().add(farm.find("@id='TESTPROJECT'")[0], 'gh:test/test#1', 'QA')
```
Comment (krzyk) [inline src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_before.groovy]: @paulodamaso why not use `project` from parameters insgtead of doing `find` (this applies to lines below also)?

File: src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_after.groovy [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import org.hamcrest.Matchers
+import org.hamcrest.collection.IsIterableContainingInAnyOrder
+import org.hamcrest.core.IsCollectionContaining
+import org.hamcrest.core.IsEqual
+import org.hamcrest.core.IsNot
+
+def exec(Project project, XML xml) {
+  Farm farm = binding.variables.farm
+  Agenda agenda = new Agenda(farm, 'test').bootstrap()
+  MatcherAssert.assertThat(
```
Comment (krzyk) [inline src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_after.groovy]: @paulodamaso what about testing that agenda of performer/DEV/REV (not inspector) is also cleared?

File: src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_before.groovy (hunk 2 above)
Comment (paulodamaso) [reply src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_before.groovy]: @krzyk `delete_stale_jobs.groovy` is executed by `PMO` project, and `PMO` projects does not accept review tasks. We need a "usual" project for testing this behavior.

File: src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_after.groovy (hunk 3 above)
Comment (paulodamaso) [reply src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_after.groovy]: @krzyk These tests should be made in `delete_stale_jobs` bundle, not here.

**Comment by paulodamaso** [comment]
@krzyk Thanks for the review, made some corrections and answered questions, please take a look.

File: src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_after.groovy (hunk 3 above)
Comment (krzyk) [reply src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_after.groovy]: @paulodamaso But I mean not the normal stale jobs, but the ones that we see in our agenda that are under QA review. They are not technically DEV jobs, but they still exist in DEV agenda **and** in QA agenda. So when a QA stale job is cleared it should also clear that job from DEV agenda.

Comment (paulodamaso) [reply src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_after.groovy]: @krzyk I disagree. If we clear these jobs from agenda it'll be painful (or even impossible) to user track jobs that are in `QA` status; actually I think the `Agenda` page relies on this to show us who is responsible for job QA.

Comment (krzyk) [reply src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_after.groovy]: @paulodamaso But we should remove them after the QA has done his/hers verdict

Comment (paulodamaso) [reply src/test/resources/com/zerocracy/bundles/delete_stale_qa_jobs/_after.groovy]: @krzyk I don't understand. 0crat is already working as expected, I don't see stale jobs in my agenda.

**Comment by krzyk** [comment]
@paulodamaso but I do see stale QA jobs in my agenda (https://github.com/zerocracy/farm/issues/1558 which was a duplicate of #1448)

**Comment by paulodamaso** [comment]
@krzyk Now I understand, sorry for all the fuss. I have a doubt about this: what defines a job as `stale` in these cases? Upon which event it should leave `DEV` agenda?

**Comment by krzyk** [comment]
@paulodamaso it should leave DEV/REV agenda at the same time as it leaves QA agenda. I think it might be already covered by your changes in delete_stale_qa_jobs,  but to make sure a test is needed

**Comment by paulodamaso** [comment]
@krzyk I've added a test assuring that it removed a stale DEV  which  job too, please take a look.

**Comment by krzyk** [comment]
@paulodamaso thanks

**Comment by krzyk** [comment]
@rultor merge

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @krzyk[/z](https://www.0crat.com/u/krzyk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1572](https://github.com/zerocracy/farm/issues/1572) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @krzyk[/z](https://www.0crat.com/u/krzyk)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**